proposals/0001.md

```diff
@@ -0,0 +1,274 @@
+* Title: *Error Handling in Rust*
+* Champion: *@FintanH, fintan.halpenny@gmail.com*
+* References:
+  * [Yoshua Wuyats error handling survey][survey]
+  * [failure][failure]
+  * [snafu][snafu]
+  * [anyhow][anyhow]
+  * [thiserror][thiserror]
+
+**Note**: The links to the packages are to versions that are at the latest as of 31/1/2019.
+
+# Introduction
+
+As good Rust developers we persevere to utilise and provide good error handling. This is a nuanced
+statement as it may cover library errors, command line errors, structured recoverable errors,
+unstructured recoverable errors, and unrecoverable errors. This proposal will examine and contrast
+the different methods and libraries that exist, ultimately making a decision as to which methods and
+library/ies our Rust developers should use when writing our libraries and applications.
+
+For a quick recap as to what's available to us out of the box see [The std](#the-std). In [Types of
+Errors](#types-of-errors) we discuss when we should use `Option`, `Result`, and what kind of errors
+we should use. In [Packages](#packages) we discuss potential libraries to use to aid error handling
+boiler-plate and help structure our own packages and applications.
+
+# Explanation
+
+## The std
+
+To ensure that we are all on the same page, it is worth covering what we mean by error handling and
+what exists for us when working with the `std` library out of the box.  When we talk about error
+handling we are referring to the possibility of failures in a function that can be handled by us,
+the author of the code.  Are we writing a function that will fail if a `Vec` is empty? Then we
+return a value that represents whether the function was successful if the `Vec` contained elements
+and failure if not.
+
+When we say "return a value", what do we mean? The two types that encapsulate a function failing are
+`Option` and `Result`. Without going through too much detail, `Option` will convey that a function
```

Comment:
What's the basis of this `Option` definition? Is the use of `Option` typically to indicate non-existence instead of failure? Example get an entry of a hashmap for a key.

Reply:
The point is to convey that `Option` should be used if the caller can infer what can go wrong. So in your example: "the key is missing". In the divide example: "the denominator is 0".

Should I make this clearer?